File: 0493-reverse-pairs/0493-reverse-pairs.cpp
class Solution {
private:
int mergeSort(vector<int>&nums,int low , int high){
    if (low >= high) return 0;
    int mid = low + (high - low) / 2;

    int count = mergeSort(nums, low, mid) + mergeSort(nums, mid + 1, high);

    int j = mid + 1;
    for(int i = low ; i <= mid ; i++){
        while(j<= high && nums[i] > 2LL*nums[j]){
            j++;
        }
        count += (j-(mid +1));
    }

    vector<int>temp;
    int left = low , right = mid +1 ;
    while(left <= mid && right <= high){
        if(nums[left] <= nums[right]){
            temp.push_back(nums[left++]);
        }
        else{
            temp.push_back(nums[right++]);
        }
    }

    while(left<= mid){
        temp.push_back(nums[left++]);
    }
    while(right <= high){
        temp.push_back(nums[right++]);
    }

    for(int i =0; i < temp.size() ; i++){
        nums[low + i] = temp[i];
    }

    return count;
}
public:
    int reversePairs(vector<int>& nums) {
        if(nums.empty()) return 0;
        return mergeSort(nums , 0 , nums.size() -1);
    }
};

File: 0493-reverse-pairs/0493-reverse-pairs.py
class Solution:
    def reversePairs(self, nums: List[int]) -> int:
        def merge_sort(low , high):
            if low >= high:
                return 0
            
            mid = (low + high) // 2

            count = merge_sort(low , mid)  + merge_sort(mid + 1 , high)

            j = mid + 1
            for i in range(low , mid + 1):
                while j <= high and nums[i] > 2*nums[j]:
                    j += 1
                count += j - (mid +1)
            
            temp = []
            left , right = low , mid + 1

            while left <= mid and right <= high:
                if nums[left] <= nums[right]:
                    temp.append(nums[left])
                    left += 1
                else:
                    temp.append(nums[right])
                    right += 1
            while left <= mid:
                temp.append(nums[left])
                left += 1
            while right <= high:
                temp.append(nums[right])
                right += 1
            
            for i in range(len(temp)):
                nums[low + i] = temp[i]
            
            return count

        return merge_sort(0,len(nums)- 1)